commandbased/commandbasedrobot.py

```diff
@@ -0,0 +1,116 @@
+import hal
+
+from wpilib.iterativerobot import IterativeRobot
+from wpilib.command.scheduler import Scheduler
+from wpilib.driverstation import DriverStation
+from wpilib.livewindow import LiveWindow
+
+
+class CommandBasedRobot(IterativeRobot):
+    '''
+    The base class for a Command-Based Robot. To use, instantiate commands and
+    trigger them.
+    '''
+
+
+    def startCompetition(self):
+        """
+        Provide an alternate "main loop" via startCompetition(). Rewritten
+        from IterativeRobot for readability and to initialize scheduler.
+        """
+
+        hal.HALReport(hal.HALUsageReporting.kResourceType_Framework,
+                      hal.HALUsageReporting.kFramework_Iterative)
+
+        self.scheduler = Scheduler.getInstance()
+        self.robotInit()
+
+        # Tell the DS that the robot is ready to be enabled
+        hal.HALNetworkCommunicationObserveUserProgramStarting()
+
+        # loop forever, calling the appropriate mode-dependent function
+        while True:
+            if self.ds.isDisabled():
+                self.disabledInit()
+                while self.ds.isDisabled():
+                    hal.HALNetworkCommunicationObserveUserProgramDisabled()
+                    self.__runPeriodic(self.disabledPeriodic)
+
+            elif self.ds.isAutonomous():
+                self.autonomousInit()
+                while self.ds.isAutonomous():
+                    hal.HALNetworkCommunicationObserveUserProgramAutonomous()
+                    self.__runPeriodic(self.autonomousPeriodic)
+
+            elif self.ds.isTest():
+                LiveWindow.setEnabled(True)
+
+                self.testInit()
+                while self.ds.isTest():
+                    hal.HALNetworkCommunicationObserveUserProgramTest()
+                    self.__runPeriodic(self.testPeriodic)
+
+                LiveWindow.setEnabled(False)
+
+            else:
+                self.teleopInit()
+                # isOperatorControl checks "not autonomous or test", so we need
+                # to check isEnabled as well, since otherwise it will continue
+                # looping while disabled.
+                while self.ds.isEnabled() and self.ds.isOperatorControl():
+                    hal.HALNetworkCommunicationObserveUserProgramTeleop()
+                    self.__runPeriodic(self.teleopPeriodic)
+
+
+    def __runPeriodic(self, periodicFunction):
+        '''
+        Run the correct "periodic" function and then wait for new data from the
+        driver station.
+        '''
+
+        periodicFunction()
+
+        while not self.ds.isNewControlData():
```

Comment:
Referring to [this line](https://github.com/robotpy/robotpy-wpilib/blob/master/wpilib/wpilib/driverstation.py#L113), I believe that guarantee can be made, as `getData` will always be called before the semaphore is notified.

It looks like they've updated their documentation in [wpilib](https://github.com/wpilibsuite/allwpilib/blob/0cd05d1a42aa7c08a212dfb40b00a28dea99efbe/wpilibc/athena/src/DriverStation.cpp#L441)... I'll file an issue to remind myself to update ours once I start on 2017 updates (November/December timeframe). 

I'll also ask for a documentation update to explicitly state this.... though, looking at their C++ implementation, it looks broken. >_>